File: Pragma_Examples/OpenMP/C/device_routine/device_routine.c
// Copyright (c) 2024 AMD HPC Application Performance Team
// Author: Bob Robey, Bob.Robey@amd.com
// MIT License

#include <stdio.h>

int main(int argc, char *argv[]){
   double ce[2]={0.0, 0.0};
#pragma target teams distribute parallel do simd reduction(+:ce1[0:2])
   for (int j = 0; j< 1000; j++){
      ce[0] += 1.0;
      ce[1] += 1.0;
   }

   printf("ce[0] = %lf ce[1] = %lf\n", ce[0], ce[1]);
   return(0);
}


#include <stdlib.h>
#include <stdio.h>

void compute(double *x);

int main(int argc, char *argv[]){
   int N=1000;
   double *x = (double *)malloc(N*sizeof(double));
#pragma omp target enter data map (alloc:x[0:N])
#pragma omp target teams distribute parallel for simd
   for (int k = 0; k < N; k++){
      x[k] = 0.0;
   }

#pragma omp target teams distribute parallel for simd
   for (int k = 0; k < N; k++){
      compute(&x[k]);
   }

   double sum = 0.0;
#pragma omp target teams distribute parallel for simd reduction(+:sum)
   for (int k = 0; k < N; k++){
      sum += x[k];
   }

   printf("Result: sum of x is %lf\n",sum);

#pragma omp target exit data map (release:x[0:N])
   free(x);
}
      

File: Pragma_Examples/OpenMP/Fortran/5_device_routine/device_routine_with_interface/2_device_routine_correct/device_routine.f90
! This example was created by Johanna Potyka
! Copyright (c) 2024 AMD HPC Application Performance Team
! MIT License

      program device_routine
      !----------------------------
      ! description: this program is meant to demonstrate 
      !              how to call a device subroutine
         use omp_lib

         implicit none

         !---variables
         integer,parameter :: N=1000
         !N                   number of values in x array
         integer,parameter :: rk=kind(1.0d0)
         !rk                  kind of real
         integer :: k, err_stat
         !k         index
         !err_stat  status variable
         real(kind=rk),dimension(:),ALLOCATABLE :: x
         !x                                        array
         real(kind=rk) :: sum
         !sum             used to sum up x
         interface

                 subroutine compute(x)
                   !$omp declare target link(compute)
                   integer,parameter :: rk=8
                   real(kind=rk), intent(inout) :: x
                 end subroutine compute
         end interface



         !---allocation
         allocate(x(1:N),STAT=err_stat)
         if(err_stat /= 0) then
             write(*,*) "error while allocating"
             STOP
         end if
         
         !---initialisation
         x = -1.0_rk
         !--- call a device subroutine in kernel
!$omp target teams distribute parallel do simd map(tofrom:x)
         do k=1,N
            call compute(x(k))
            !x(k) = 1.0_rk
         end do
!$omp end target teams distribute parallel do simd         

         !--- initialize sum
        sum = 0.0_rk;

        !--- sum up x to sum on device with reduction
!$omp target teams distribute parallel do simd reduction(+:sum) map(to:x)
        do k=1,N
           sum = sum + x(k)
        end do
!$omp end target teams distribute parallel do simd        

        !--- print result
        Write(*,'(A,F0.12)') "Result: sum of x is ",sum

        call compute(x(1))
       deallocate(x) 

      end program device_routine
      
